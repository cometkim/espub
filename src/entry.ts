import * as path from 'node:path';
import type { ConditionalExport, Config } from './config';
import type { Reporter } from './report';

export type Entry = {
  key: string,
  entryPath: string,
  mode: 'development' | 'production',
  platform: 'netural' | 'browser' | 'deno' | 'node',
  module: 'commonjs' | 'esmodule' | 'dts' | 'file',
  sourceFile: string[],
  outputFile: string,
};

interface GetEntriesFromConfig {
  (props: {
    config: Config,
    rootDir: string,
    outDir: string,
    resolvePath: (path: string) => string,
    reporter: Reporter,
  }): Entry[];
};

export const getEntriesFromConfig: GetEntriesFromConfig = ({
  config,
  rootDir,
  outDir,
  reporter,
  resolvePath,
}) => {
  const resolvedOutDir = resolvePath(outDir);
  const resolvedRootDir = resolvePath(rootDir);

  const defaultPlatform: Entry['platform'] = 'netural';
  const defaultMode: Entry['mode'] = 'production';
  const defaultModule: Entry['module'] = (
    (config.type === 'module')
      ? 'esmodule'
      : 'commonjs'
  );

  const entryMap = new Map<Entry['entryPath'], Entry>();

  function addEntry({
    key, entryPath, platform, module, mode,
  }: {
    key: string,
    entryPath: string,
    platform: Entry['platform'],
    mode: Entry['mode'],
    module: Entry['module'],
  }) {
    if (!entryPath.startsWith('./')) {
      reporter.error(`Invalid entry "${key}", entry path should starts with \`"./"\``);
      throw new Error('FIXME');
    }

    if (key.includes('*') || entryPath.includes('*')) {
      reporter.warn(`Ignoring ${key}: subpath pattern(\`*\`) is not supported yet`);
      return;
    }

    const entry = entryMap.get(entryPath);
    if (entry && (entry.platform !== platform || entry.module !== module)) {
      reporter.warn(`
Conflict found for ${path}

  ${entry.key}
  { module: "${entry.module}", platform: "${entry.platform}" }

  vs

  (ignored) ${key}
  { module: "${module}", platform: "${platform}" }

Did you forget to specify the Node.js version in the "engines" field?
Or you may not need to specify "require" or "node" entries.
`);
      return;
    }

    const resolvedEntryPath = resolvePath(entryPath);
    entryMap.set(entryPath, {
      key,
      entryPath,
      mode,
      platform,
      module,
      sourceFile: [
        resolvedEntryPath.replace(
          resolvedOutDir,
          resolvedRootDir,
        ),
      ],
      outputFile: resolvedEntryPath,
    });
  }

  function addMainEntry({
    key, entryPath,
  }: {
    key: string,
    entryPath: string,
  }) {
    const ext = path.extname(entryPath);
    switch (ext) {
      case '.cjs': {
        addEntry({
          key,
          platform: defaultPlatform,
          mode: defaultMode,
          module: 'commonjs',
          entryPath,
        });
        break;
      }
      case '.mjs': {
        addEntry({
          key,
          platform: defaultPlatform,
          mode: defaultMode,
          module: 'esmodule',
          entryPath,
        });
        break;
      }
      case '.node': {
        addEntry({
          key,
          platform: 'node',
          mode: defaultMode,
          module: 'file',
          entryPath,
        });
        break;
      }
      case '.json': {
        addEntry({
          key,
          platform: defaultPlatform,
          mode: defaultMode,
          module: 'file',
          entryPath,
        });
        break;
      }
      // case '.js':
      default: {
        addEntry({
          key,
          platform: defaultPlatform,
          mode: defaultMode,
          module: defaultModule,
          entryPath,
        });
        break;
      }
    }
  }

  function addModuleEntry({
    key, entryPath,
  }: {
    key: string,
    entryPath: string,
  }) {
    const ext = path.extname(entryPath);
    if (ext === '.js' || ext === '.mjs') {
      addEntry({
        key,
        platform: defaultPlatform,
        mode: defaultMode,
        module: 'esmodule',
        entryPath,
      });
    } else {
      // FIXME: warn
    }
  }

  function addConditionalEntry({
    key, parentKey, platform, mode, module, entryPath,
  }: {
    key: string,
    parentKey: string | null,
    platform: Entry['platform'],
    mode: Entry['mode'],
    module: Entry['module'],
    entryPath: ConditionalExport,
  }) {
    if (typeof entryPath === 'string') {
      addEntry({
        key,
        mode,
        entryPath,
      });
    } else if (typeof entryPath === 'object') {
      for (const [currentKey, output] of Object.entries(entryPath)) {
        // See https://nodejs.org/api/packages.html#packages_community_conditions_definitions
        switch (currentKey) {
          case 'import': {
            addConditionalEntry({
              key: `${parentKey}.import`,
              parentKey: 'import',
              platform,
              mode,
              module: 'esmodule',
              entryPath: output,
            });
          }
          case 'require': {
            addConditionalEntry({
              key: `${parentKey}.require`,
              parentKey: 'require',
              platform,
              mode,
              module: 'commonjs',
              entryPath: output,
            });
          }
          case 'types': {
            addConditionalEntry({
              key: `${parentKey}.types`,
              parentKey: 'types',
              platform,
              mode,
              module: 'dts',
              entryPath: output,
            });
          }
          case 'deno': {
            addConditionalEntry({
              key: `${parentKey}.deno`,
              parentKey: 'deno',
              platform: 'deno',
              mode,
              module,
              entryPath: output,
            });
          }
          case 'browser': {
            addConditionalEntry({
              key: `${parentKey}.browser`,
              parentKey: 'browser',
              platform: 'browser',
              mode,
              module,
              entryPath: output,
            });
          }
          case 'development': {
            addConditionalEntry({
              key: `${parentKey}.development`,
              parentKey: 'development',
              platform,
              mode: 'development',
              module,
              entryPath: output,
            });
          }
          case 'production': {
            addConditionalEntry({
              key: `${parentKey}.production`,
              parentKey: 'production',
              platform,
              mode: 'production',
              module,
              entryPath: output,
            });
          }
          default: {
          }
        }
      }
    }
  }

  if (config.exports) {
    addConditionalEntry({
      key: 'exports',
      parentKey: null,
      platform: defaultPlatform,
      mode: 'production',
      module: defaultModule,
      entryPath: config.exports,
    });
  } else {
    reporter.warn(`Using "exports" field is highly recommended.
See https://nodejs.org/api/packages.html for more detail.
`);
  }

  if (typeof config.main === 'string') {
    addMainEntry({
      key: 'main',
      entryPath: config.main,
    });
  }

  if (typeof config.module === 'string') {
    addModuleEntry({
      key: 'module',
      entryPath: config.module,
    });

    reporter.warn(`"module" field is not standard and may works in only legacy bundlers. Consider using "exports" instead.
See https://nodejs.org/api/packages.html for more detail.
`);
  }

  return Array.from(entryMap.values());
}
