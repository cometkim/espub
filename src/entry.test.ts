import * as path from 'node:path';
import { describe, expect, vi } from 'vitest';

import type { Config } from './config';
import type { Reporter } from './report';
import type { Entry } from './entry';
import { getEntriesFromConfig } from './entry';

describe('getEntriesFromConfig', test => {
  const resolvePath = (to: string) => path.join('/project', to);

  const reporter: Reporter = {
    debug: vi.fn(),
    info: vi.fn(),
    warn: vi.fn(),
    error: vi.fn(),
  };

  const entriesFromConfig = (config: Config) => {
    return getEntriesFromConfig({
      config,
      resolvePath,
      reporter,
      rootDir: './src',
      outDir: './lib',
    });
  };

  test('empty package', () => {
    expect(entriesFromConfig({
      name: 'my-package',
    })).toEqual([]);
  });

  test('main entry, implicit commonjs', () => {
    expect(entriesFromConfig({
      name: 'my-package',
      main: './lib/index.js',
    })).toEqual<Entry[]>([
      {
        key: 'main',
        module: 'commonjs',
        mode: 'production',
        platform: 'netural',
        entryPath: './lib/index.js',
        sourceFile: [
          '/project/src/index.cjs',
          '/project/src/index.js',
        ],
        outputFile: '/project/lib/index.js',
      },
    ]);
  });

  test('main field, explicit commonjs', () => {
    expect(entriesFromConfig({
      name: 'my-package',
      type: 'commonjs',
      main: './lib/index.js',
    })).toEqual<Entry[]>([
      {
        key: 'main',
        module: 'commonjs',
        mode: 'production',
        platform: 'netural',
        entryPath: './lib/index.js',
        sourceFile: [
          '/project/src/index.cjs',
          '/project/src/index.js',
        ],
        outputFile: '/project/lib/index.js',
      },
    ]);
  });

  test('main field, explicit esmodule', () => {
    expect(entriesFromConfig({
      name: 'my-package',
      type: 'module',
      main: './lib/index.js',
    })).toEqual<Entry[]>([
      {
        key: 'main',
        module: 'esmodule',
        mode: 'production',
        platform: 'netural',
        entryPath: './lib/index.js',
        sourceFile: [
          '/project/src/index.mjs',
          '/project/src/index.js',
        ],
        outputFile: '/project/lib/index.js',
      },
    ]);
  });

  test('main field with module extension', () => {
    expect(entriesFromConfig({
      name: 'my-package',
      type: 'module',
      main: './lib/index.cjs',
    })).toEqual<Entry[]>([
      {
        key: 'main',
        module: 'commonjs',
        mode: 'production',
        platform: 'netural',
        entryPath: './lib/index.cjs',
        sourceFile: [
          '/project/src/index.cjs',
          '/project/src/index.js',
        ],
        outputFile: '/project/lib/index.cjs',
      },
    ]);

    expect(entriesFromConfig({
      name: 'my-package',
      type: 'commonjs',
      main: './lib/index.mjs',
    })).toEqual<Entry[]>([
      {
        key: 'main',
        module: 'esmodule',
        mode: 'production',
        platform: 'netural',
        entryPath: './lib/index.mjs',
        sourceFile: [
          '/project/src/index.mjs',
          '/project/src/index.js',
        ],
        outputFile: '/project/lib/index.mjs',
      },
    ]);
  });

  test('main field accepts js, json, node addon entryPaths', () => {
    expect(entriesFromConfig({
      name: 'my-package',
      main: './lib/index.json',
    })).toEqual<Entry[]>([
      {
        key: 'main',
        module: 'file',
        mode: 'production',
        platform: 'netural',
        entryPath: './lib/index.json',
        sourceFile: [
          '/project/src/index.json',
        ],
        outputFile: '/project/lib/index.json',
      },
    ]);

    expect(entriesFromConfig({
      name: 'my-package',
      main: './lib/index.node',
    })).toEqual<Entry[]>([
      {
        key: 'main',
        module: 'file',
        mode: 'production',
        entryPath: './lib/index.node',
        platform: 'node',
        sourceFile: [
          '/project/src/index.node',
        ],
        outputFile: '/project/lib/index.node',
      },
    ]);

    // otherwise treated as JavaScript text
    expect(entriesFromConfig({
      name: 'my-package',
      main: './lib/index.css',
    })).toEqual<Entry[]>([
      {
        key: 'main',
        module: 'commonjs',
        mode: 'production',
        entryPath: './lib/index.css',
        platform: 'netural',
        sourceFile: [
          '/project/src/index.css.cjs',
          '/project/src/index.css.js',
          '/project/src/index.css',
        ],
        outputFile: '/project/lib/index.css',
      },
    ]);
  });

  test('module field', () => {
    expect(entriesFromConfig({
      name: 'my-package',
      main: './lib/main.js',
      module: './lib/module.js',
    })).toEqual<Entry[]>([
      {
        key: 'main',
        module: 'commonjs',
        mode: 'production',
        platform: 'netural',
        entryPath: './lib/main.js',
        sourceFile: [
          '/project/src/main.cjs',
          '/project/src/main.js',
        ],
        outputFile: '/project/lib/main.js',
      },
      {
        key: 'module',
        module: 'esmodule',
        mode: 'production',
        platform: 'netural',
        entryPath: './lib/module.js',
        sourceFile: [
          '/project/src/module.mjs',
          '/project/src/module.js',
        ],
        outputFile: '/project/lib/module.js',
      },
    ]);
  });

  test('prefer "main" over "module" on conflict', () => {
    expect(entriesFromConfig({
      name: 'my-package',
      main: './lib/index.js',
      module: './lib/index.js',
    })).toEqual<Entry[]>([
      {
        key: 'main',
        module: 'commonjs',
        mode: 'production',
        platform: 'netural',
        entryPath: './lib/index.js',
        sourceFile: [
          '/project/src/index.cjs',
          '/project/src/index.js',
        ],
        outputFile: '/project/lib/index.js',
      },
    ]);

    expect(entriesFromConfig({
      name: 'my-package',
      module: './lib/index.js',
      main: './lib/index.js',
    })).toEqual<Entry[]>([
      {
        key: 'main',
        module: 'commonjs',
        mode: 'production',
        platform: 'netural',
        entryPath: './lib/index.js',
        sourceFile: [
          '/project/src/index.cjs',
          '/project/src/index.js',
        ],
        outputFile: '/project/lib/index.js',
      },
    ]);
  });

  test('prefer "exports" over "module" on conflict', () => {
    expect(entriesFromConfig({
      name: 'my-package',
      exports: './lib/index.js',
      module: './lib/index.js',
    })).toEqual<Entry[]>([
      {
        key: 'exports',
        module: 'commonjs',
        mode: 'production',
        platform: 'netural',
        entryPath: './lib/index.js',
        sourceFile: [
          '/project/src/index.cjs',
          '/project/src/index.js',
        ],
        outputFile: '/project/lib/index.js',
      },
    ]);

    expect(entriesFromConfig({
      name: 'my-package',
      module: './lib/index.js',
      exports: './lib/index.js',
    })).toEqual<Entry[]>([
      {
        key: 'exports',
        module: 'commonjs',
        mode: 'production',
        platform: 'netural',
        entryPath: './lib/index.js',
        sourceFile: [
          '/project/src/index.cjs',
          '/project/src/index.js',
        ],
        outputFile: '/project/lib/index.js',
      },
    ]);
  });

  test('exports field', () => {
    expect(entriesFromConfig({
      name: 'my-package',
      exports: './lib/index.js',
    })).toEqual<Entry[]>([
      {
        key: 'exports',
        module: 'commonjs',
        mode: 'production',
        platform: 'netural',
        entryPath: './lib/index.js',
        sourceFile: [
          '/project/src/index.cjs',
          '/project/src/index.js',
        ],
        outputFile: '/project/lib/index.js',
      },
    ]);
  });

  test('exports field always precedense over main', () => {
    expect(entriesFromConfig({
      name: 'my-package',
      main: './lib/index.js',
      exports: './lib/index.js',
    })).toEqual<Entry[]>([
      {
        key: 'exports',
        module: 'commonjs',
        mode: 'production',
        platform: 'netural',
        entryPath: './lib/index.js',
        sourceFile: [
          '/project/src/index.cjs',
          '/project/src/index.js',
        ],
        outputFile: '/project/lib/index.js',
      },
    ]);

    expect(entriesFromConfig({
      name: 'my-package',
      type: 'module',
      main: './lib/index.js',
      exports: {
        require: './lib/index.js',
      },
    })).toEqual<Entry[]>([
      {
        key: 'exports',
        module: 'commonjs',
        mode: 'production',
        platform: 'netural',
        entryPath: './lib/index.js',
        sourceFile: [
          '/project/src/index.cjs',
          '/project/src/index.js',
        ],
        outputFile: '/project/lib/index.js',
      },
    ]);
  });

  test('conditional exports', () => {
    expect(entriesFromConfig({
      name: 'my-package',
      exports: {
        require: './lib/require.js',
        import: './lib/import.js',
      },
    })).toEqual<Entry[]>([
      {
        key: 'exports.require',
        module: 'commonjs',
        mode: 'production',
        platform: 'netural',
        entryPath: './lib/require.js',
        sourceFile: [
          '/project/src/require.cjs',
          '/project/src/require.js',
        ],
        outputFile: '/project/lib/require.js',
      },
      {
        key: 'exports.import',
        module: 'esmodule',
        mode: 'production',
        platform: 'netural',
        entryPath: './lib/import.js',
        sourceFile: [
          '/project/src/import.mjs',
          '/project/src/import.js',
        ],
        outputFile: '/project/lib/import.js',
      },
    ]);

    expect(entriesFromConfig({
      name: 'my-package',
      exports: {
        '.': './lib/index.js',
        './index': './lib/index.js',
        './index.js': './lib/index.js',
        './feature': './lib/feature/index.js',
        './feature/index.js': './lib/feature/index.js',
        './package.json': './package.json',
      },
    })).toEqual<Entry[]>([
      {
        key: 'exports["."]',
        module: 'commonjs',
        mode: 'production',
        platform: 'netural',
        entryPath: './lib/index.js',
        sourceFile: [
          '/project/src/index.cjs',
          '/project/src/index.js',
        ],
        outputFile: '/project/lib/index.js',
      },
      {
        key: 'exports["./feature"]',
        module: 'commonjs',
        mode: 'production',
        platform: 'netural',
        entryPath: './feature/index.js',
        sourceFile: [
          '/project/src/feature/index.cjs',
          '/project/src/feature/index.js',
        ],
        outputFile: '/project/lib/feature/index.js',
      },
    ]);
  });

  test('nested conditional exports', () => {
    expect(entriesFromConfig({
      name: 'my-package',
      exports: {
        '.': {
          require: './lib/index.cjs',
          import: './lib/index.js',
        },
      },
    })).toEqual<Entry[]>([
      {
        key: 'exports["."].require',
        module: 'commonjs',
        mode: 'production',
        platform: 'netural',
        entryPath: './lib/index.cjs',
        sourceFile: [
          '/project/src/index.cjs',
          '/project/src/index.js',
        ],
        outputFile: '/project/lib/index.cjs',
      },
      {
        key: 'exports["."].import',
        module: 'esmodule',
        mode: 'production',
        platform: 'netural',
        entryPath: './lib/index.js',
        sourceFile: [
          '/project/src/index.mjs',
          '/project/src/index.js',
        ],
        outputFile: '/project/lib/index.js',
      },
    ]);
  });

  test('conditional exports community definitions', () => {
    expect(entriesFromConfig({
      name: 'my-package',
      exports: {
        '.': {
          node: {
            require: './lib/index.cjs',
            import: {
              development: './lib/index.mjs',
              production: './lib/index.min.mjs',
            },
          },
          browser: {
            development: './lib/browser.js',
            production: './lib/browser.min.js',
          },
        },

        // reversed exports map
        deno: {
          './deno': './lib/deno.js',
        },
      },
    })).toEqual<Entry[]>([
      {
        key: 'exports["."].node.require',
        module: 'commonjs',
        mode: 'production',
        platform: 'node',
        entryPath: './lib/index.cjs',
        sourceFile: [
          '/project/src/index.cjs',
          '/project/src/index.js',
        ],
        outputFile: '/project/lib/index.cjs',
      },
      {
        key: 'exports["."].node.import.development',
        module: 'esmodule',
        mode: 'development',
        platform: 'node',
        entryPath: './lib/index.mjs',
        sourceFile: [
          '/project/src/index.mjs',
          '/project/src/index.js',
        ],
        outputFile: '/project/lib/index.mjs',
      },
      {
        key: 'exports["."].node.import.production',
        module: 'esmodule',
        mode: 'production',
        platform: 'node',
        entryPath: './lib/index.min.mjs',
        sourceFile: [
          '/project/src/index.mjs',
          '/project/src/index.js',
        ],
        outputFile: '/project/lib/index.min.mjs',
      },

      // Browsers don't support CommonJS.
      // However, Node.js tool users may have their own bundler.
      //
      // It shouldn't be recommended
      {
        key: 'exports["."].browser.development',

        module: 'commonjs',
        mode: 'development',
        platform: 'browser',
        entryPath: './lib/browser.js',
        sourceFile: [
          '/project/src/browser.mjs',
          '/project/src/browser.js',
        ],
        outputFile: '/project/lib/browser.js',
      },

      // Using some popular extension patterns like `*.min.js` can be inferred as regular extension.
      {
        key: 'exports["."].browser.production',
        module: 'commonjs',
        mode: 'production',
        platform: 'browser',
        entryPath: './lib/browser.min.js',
        sourceFile: [
          '/project/src/browser.mjs',
          '/project/src/browser.js',
        ],
        outputFile: '/project/lib/browser.min.js',
      },

      // Deno only supports ESM.
      // However, Deno users can load CommonJS module using compat API (https://deno.land/std/node/module.ts) if they really want.
      // See https://deno.land/manual/npm_nodejs/std_node
      //
      // It shouldn't be recommended anyway
      {
        key: 'exports.deno["./deno"]',
        module: 'commonjs',
        mode: 'production',
        platform: 'deno',
        entryPath: './lib/deno.js',
        sourceFile: [
          '/project/src/deno.cjs',
          '/project/src/deno.js',
        ],
        outputFile: '/project/lib/deno.js',
      },
    ]);
  });

  describe.todo('in TypeScript project');

  describe.todo('when rootDir=./src, outDir=.');
  describe.todo('when rootDir=., outDir=./lib');
  describe.todo('when rootDir=outDir=.');
  describe.todo('when rootDir=outDir=./lib');
  describe.todo('when rootDir=outDir=./lib & TypeScript sources');
});
