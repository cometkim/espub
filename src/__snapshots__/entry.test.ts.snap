// Vitest Snapshot v1

exports[`getEntriesFromContext > bin entry only accepts js 1`] = `
"Only JavaScript files are allowed for bin entry.
"
`;

exports[`getEntriesFromContext > throw if "main" and "module" is on conflict 1`] = `
"Hint: Did you forgot to set \\"type\\" to 'module' for ESM-first approach?

"
`;

exports[`getEntriesFromContext - in TypeScript project > types entry does not accept nesting 1`] = `
"\\"types\\" entry must be .d.ts file and cannot be nested!
"
`;

exports[`getEntriesFromContext - in TypeScript project > types entry must has .d.ts extension 1`] = `
"Only .d.ts or .d.cts or .d.mts allowed for \\"types\\" entry.
"
`;

exports[`getEntriesFromContext - in TypeScript project > types entry must occur first in conditional exports 1`] = `
"\\"types\\" entry must occur first in conditional exports for correct type resolution.
"
`;
